plutus-tx-plugin/executables/profile/Main.hs

```diff
@@ -24,12 +42,120 @@ import qualified PlutusCore.Default        as PLC
 
 import           Control.Lens.Combinators  (_2)
 import           Control.Lens.Getter       (view)
+import           Data.List                 (intercalate)
+import           Data.Maybe                (fromJust)
 import           Data.Proxy                (Proxy (Proxy))
-import           Data.Text                 (Text)
+import qualified Data.Text                 as T
+import           Data.Time.Clock           (NominalDiffTime, UTCTime, diffUTCTime)
 import           Prettyprinter.Internal    (pretty)
 import           Prettyprinter.Render.Text (hPutDoc)
 import           System.IO                 (IOMode (WriteMode), withFile)
 
+data StackFrame
+  = MkStackFrame
+  { -- | The variable name.
+    varName           :: T.Text,
+    -- | The time when it starts to be evaluated.
+    startTime         :: UTCTime,
+    -- | The time spent on evaluating the functions it called.
+    timeSpentCalledFn :: NominalDiffTime
+  }
+  deriving (Show)
+
+-- | Write the time log of a program to a file in
+-- the plutus-tx-plugin/executables/profile/ directory.
+writeLogToFile ::
+  ToUPlc a PLC.DefaultUni PLC.DefaultFun =>
+  -- | Name of the file you want to save it as.
+  FilePath ->
+  -- | The program to be profiled.
+  [a] ->
+  IO ()
+writeLogToFile fileName values = do
+  let filePath = "plutus-tx-plugin/executables/profile/"<>fileName<>".timelog"
+  log <- T.intercalate "\n" . view _2 <$> (rethrow $ runUPlcProfileExec values)
+  writeFile filePath (T.unpack log)
+  processed <- processLog filePath
+  writeFile (filePath<>".stacks") (printProcessedLog processed)
+  pure ()
+
+data ProfileEvent =
+  MkProfileEvent UTCTime Transition T.Text
+
+data Transition =
+  Enter
+  | Exit
+
+data StackTime =
+  MkStackTime [T.Text] Double
+
+processLog :: FilePath -> IO [StackTime]
+processLog file = do
+  content <- readFile file
+  let lEvents = lines content -- turn to a list of events
+  pure $ getStacks (profileEvents lEvents)
+
+printProcessedLog :: [StackTime] -> String
+printProcessedLog ((MkStackTime fns duration):tl) =
+    intercalate
+      "; "
+      -- reverse to make the functions in the order correct for flamegraphs.
+      (reverse (map T.unpack fns))
+      <>" "
+      -- multiplying the duration by a large number
+      -- to make the number show in a way flamegraph accepts.
```

Comment:
Hmmm. Better to make the time field of `StackTime` an integer or something? Indeed, perhaps this is an argument that we shouldn't print UTCTime but rather UnixTime or something that just looks like an integer.

Reply:
It would be nice if this was somewhat meaningful, although mostly people are looking at the relative sizes.

Reply:
Thanks! Looking at the UnixTime package I realized I could use `nominalDiffTimeToSeconds` to turn it into seconds. It's showing it in decimals so flamegraph accepts it fine.

Reply:
Hm...although it does draw the graph without errors, when you hover the numbers are always 0.00% because of the small number of seconds. So I will turn it into microseconds (like in UnixTime) for readability.